daemon: Clean endpoint BPF map on start

vadorovsky opened the pull request:
Call DeleteAll method of LXCMap on every daemon start, because
some previously existing endpoints and IP addresses can be not
valid anymore (i.e. when we change pod CIDR before restarting
cilium).

Fixes #2748

Signed-off-by: Michal Rostecki <mrostecki@suse.com>

```release-note
Clean endpoint BPF map on daemon start
```

vadorovsky committed:
daemon: Clean endpoint BPF map on start

Call DeleteAll method of LXCMap on every daemon start, because
some previously existing endpoints and IP addresses can be not
valid anymore (i.e. when we change pod CIDR before restarting
cilium).

Fixes #2748

Signed-off-by: Michal Rostecki <mrostecki@suse.com>

aanm commented on pkg/endpointmanager/conntrack.go:
```diff
@@ -152,6 +153,51 @@ func ResetProxyPort(ipv4Enabled bool, e *endpoint.Endpoint, isLocal bool, ips []
+		for _, keyToPreserve := range ep.GetBPFKeys() {
+			for i, key := range epsi.keys {
+				if key == keyToPreserve {
+					epsi.keys = append(epsi.keys[:i], epsi.keys[i+1:]...)
```
This will delete all endpoints that do not belong to the list of endpoints running

aanm reviewed (changes requested):
I understand what you tried to achieve.
Unfortunately not all endpoints stored in the bpf map belong to endpoints.

For example: this is the output of a running agent with all the addresses that should be kept:
```
IP ADDRESS           LOCAL ENDPOINT INFO
10.0.106.144         id=28441 ifindex=21  mac=00:00:00:00:B2:D6 nodemac=EF:86:00:00:D6:C9 seclabel=0x2146 portMaps=(empty)
10.0.131.27          id=17607 ifindex=25  mac=00:00:00:00:02:C9 nodemac=58:E3:00:00:16:76 seclabel=0x4490 portMaps=(empty)
10.0.195.104         id=17990 ifindex=23  mac=00:00:00:00:12:F5 nodemac=3D:5C:00:00:C2:12 seclabel=0x692e portMaps=(empty)
10.0.213.77          id=17471 ifindex=43  mac=00:00:00:00:BE:75 nodemac=1A:4C:00:00:BE:0D seclabel=0xf67f portMaps=(empty)
10.0.242.54          id=29898 ifindex=13  mac=00:00:00:00:DA:65 nodemac=6A:0F:00:00:D6:4D seclabel=0xdbba portMaps=(empty)
10.0.244.141         id=50640 ifindex=41  mac=00:00:00:00:CA:18 nodemac=86:CF:00:00:CA:51 seclabel=0x653f portMaps=(empty)
10.0.28.238          (localhost)
10.0.54.125          id=11004 ifindex=17  mac=00:00:00:00:8A:CD nodemac=A7:AB:00:00:46:F3 seclabel=0x43c portMaps=(empty)
10.0.65.87           id=13233 ifindex=35  mac=00:00:00:00:8E:35 nodemac=5A:E3:00:00:B6:0E seclabel=0x67bc portMaps=(empty)
192.168.33.11        (localhost)
f00d::a00:0:0:2afc   id=11004 ifindex=17  mac=00:00:00:00:8A:CD nodemac=A7:AB:00:00:46:F3 seclabel=0x43c portMaps=(empty)
f00d::a00:0:0:33b1   id=13233 ifindex=35  mac=00:00:00:00:8E:35 nodemac=5A:E3:00:00:B6:0E seclabel=0x67bc portMaps=(empty)
f00d::a00:0:0:443f   id=17471 ifindex=43  mac=00:00:00:00:BE:75 nodemac=1A:4C:00:00:BE:0D seclabel=0xf67f portMaps=(empty)
f00d::a00:0:0:44c7   id=17607 ifindex=25  mac=00:00:00:00:02:C9 nodemac=58:E3:00:00:16:76 seclabel=0x4490 portMaps=(empty)
f00d::a00:0:0:4646   id=17990 ifindex=23  mac=00:00:00:00:12:F5 nodemac=3D:5C:00:00:C2:12 seclabel=0x692e portMaps=(empty)
f00d::a00:0:0:6f19   id=28441 ifindex=21  mac=00:00:00:00:B2:D6 nodemac=EF:86:00:00:D6:C9 seclabel=0x2146 portMaps=(empty)
f00d::a00:0:0:74ca   id=29898 ifindex=13  mac=00:00:00:00:DA:65 nodemac=6A:0F:00:00:D6:4D seclabel=0xdbba portMaps=(empty)
f00d::a00:0:0:8ad6   (localhost)
f00d::a00:0:0:c5d0   id=50640 ifindex=41  mac=00:00:00:00:CA:18 nodemac=86:CF:00:00:CA:51 seclabel=0x653f portMaps=(empty)
fd01::b              (localhost)
```

On this case, all of the `(localhost)` addresses should be kept. The issue #2748 is because some of the `(localhost)` IP addresses are no longer valid, (e.g. running cilium with a different pod-cidr than the current one, which can cause the datapath to think remote endpoints are local endpoints)

I think it will be better to blindly delete all endpoints in the BPF map when starting cilium and the restoring functionality will take care of filling the missing endpoints into the BPF map.

vadorovsky committed:
pkg/maps/lxcmap: Add DeleteAll function

The exported function which allows the other modules to clean
the endpoint map.

Signed-off-by: Michal Rostecki <mrostecki@suse.com>
